Support covariant return types for overriding functions

### Describe the project you are working on

I am working on an addon that manages RPG data.

### Describe the problem or limitation you are having in your project

It is currently impossible to specify a more 'specific' type of a return value of any overriden function, if that returned value is a subtype of the overriding function. Many common languages like Java and ~C#~ C++ support such feature ([see table](https://github.com/godotengine/godot-proposals/issues/7450#issuecomment-1666951033)), as it unlocks additional functionality for polymorphism.

Currently, Godot 4.1 produces an error like this:

<img width="804" alt="Screenshot 2023-08-05 151230" src="https://github.com/godotengine/godot-proposals/assets/822035/dc75dedd-928c-48d5-a585-603d9f8b8345">


### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Support covariant return types on functions if the return value is a sub-type of the overriding function's return value.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

This script does not compile as part of Godot 4.1 but after this proposal was implemented it would work as expected:
```gdscript
class BaseClass extends RefCounted:

	func greetings() -> void:
		print("Hello!")

class SubClass extends BaseClass:
	
	func greetings() -> void:
		print("Hello Sub Class!")

class BaseGreeting extends RefCounted:
	
	func get_obj() -> BaseClass:
		return BaseClass.new()

class SubGreeting extends BaseGreeting:
	
	func get_obj() -> SubClass:
		return SubClass.new()

func _ready():
	var base:BaseGreeting = SubGreeting.new()
	var obj = base.get_obj()
	obj.greetings() # should print 'Hello Sub Class!'
```
Example project of the above:

[Covariant-Example.zip](https://github.com/godotengine/godot-proposals/files/12267448/Covariant-Example.zip)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It can't since this is a GDScript core language feature that cannot be worked around easily.

### Is there a reason why this should be core and not an add-on in the asset library?

Cannot be done via addons nor GDExtension.

I did some tests for some context.

| Language | Test | Behavior |
| :--: | :--: | :--: |
| Java | [Test](https://replit.com/@adamscott/SupportCovariantReturnJava) | Behaves as the proposal suggests |
| Swift | [Test](https://replit.com/@adamscott/SupportCovariantReturnSwift) | Behaves as the proposal suggests |
| Typescript/Javascript |[Test](https://replit.com/@adamscott/SupportCovariantReturnTypescript)| Behaves as the proposal suggests <br> _*it don't really checks for signature changes_ |
| C# | [Test](https://replit.com/@adamscott/SupportCovariantReturnCSharp) | Fails. <br> `error CS0115: 'SubGreeting.greetings()': no suitable method found to override` <br> It cannot override the `greetings()` function as the parent class doesn't have the signature `virtual public SubClass greetings()` |
| C++ | [Test](https://replit.com/@adamscott/SupportCovariantReturnCPlusPlus) | Behaves as the proposal suggests |
| OCaml | [Test](https://replit.com/@adamscott/SupportCovariantReturnOCaml) | Behaves as the proposal suggests <br> _*it don't really checks for signature changes_ |

--- a/modules/gdscript/gdscript_analyzer.cpp
+++ b/modules/gdscript/gdscript_analyzer.cpp
@@ -1671,15 +1671,42 @@ void GDScriptAnalyzer::resolve_function_signature(GDScriptParser::FunctionNode *
 		StringName native_base;
 		if (!p_is_lambda && get_function_signature(p_function, false, base_type, function_name, parent_return_type, parameters_types, default_par_count, method_flags, &native_base)) {
 			bool valid = p_function->is_static == method_flags.has_flag(METHOD_FLAG_STATIC);
-			valid = valid && parent_return_type == p_function->get_datatype();
+
+			if (p_function->return_type != nullptr) {
+				// Check return type covariance.
+				GDScriptParser::DataType return_type = p_function->get_datatype();
+				if (return_type.is_variant()) {
+					// `is_type_compatible()` returns `true` if one of the types is `Variant`.
+					// Don't allow an explicitly specified `Variant` if the parent return type is narrower.
+					valid = valid && parent_return_type.is_variant();
+				} else if (return_type.kind == GDScriptParser::DataType::BUILTIN && return_type.builtin_type == Variant::NIL) {
+					// `is_type_compatible()` returns `true` if target is an `Object` and source is `null`.
+					// Don't allow `void` if the parent return type is a hard non-`void` type.
+					if (parent_return_type.is_hard_type() && !(parent_return_type.kind == GDScriptParser::DataType::BUILTIN && parent_return_type.builtin_type == Variant::NIL)) {
+						valid = false;
+					}
+				} else {
+					valid = valid && is_type_compatible(parent_return_type, return_type);
+				}
+			}
 
 			int par_count_diff = p_function->parameters.size() - parameters_types.size();
 			valid = valid && par_count_diff >= 0;
 			valid = valid && default_value_count >= default_par_count + par_count_diff;
 
-			int i = 0;
-			for (const GDScriptParser::DataType &par_type : parameters_types) {
-				valid = valid && par_type == p_function->parameters[i++]->get_datatype();
+			if (valid) {
+				int i = 0;
+				for (const GDScriptParser::DataType &parent_par_type : parameters_types) {
+					// Check parameter type contravariance.
+					GDScriptParser::DataType current_par_type = p_function->parameters[i++]->get_datatype();
+					if (parent_par_type.is_variant() && parent_par_type.is_hard_type()) {
+						// `is_type_compatible()` returns `true` if one of the types is `Variant`.
+						// Don't allow narrowing a hard `Variant`.
+						valid = valid && current_par_type.is_variant();
+					} else {
+						valid = valid && is_type_compatible(current_par_type, parent_par_type);
+					}
+				}
 			}
 
 			if (!valid) {

--- a/modules/gdscript/tests/scripts/analyzer/errors/function_param_type_invalid_contravariance_1.gd
+++ b/modules/gdscript/tests/scripts/analyzer/errors/function_param_type_invalid_contravariance_1.gd
@@ -1,0 +1,10 @@
+class A:
+	func f(_p: Object):
+		pass
+
+class B extends A:
+	func f(_p: Node):
+		pass
+
+func test():
+	pass

--- a/modules/gdscript/tests/scripts/analyzer/errors/function_param_type_invalid_contravariance_1.out
+++ b/modules/gdscript/tests/scripts/analyzer/errors/function_param_type_invalid_contravariance_1.out
@@ -1,0 +1,2 @@
+GDTEST_ANALYZER_ERROR
+The function signature doesn't match the parent. Parent signature is "f(Object) -> Variant".

--- a/modules/gdscript/tests/scripts/analyzer/errors/function_param_type_invalid_contravariance_2.gd
+++ b/modules/gdscript/tests/scripts/analyzer/errors/function_param_type_invalid_contravariance_2.gd
@@ -1,0 +1,10 @@
+class A:
+	func f(_p: Variant):
+		pass
+
+class B extends A:
+	func f(_p: Node): # No `is_type_compatible()` misuse.
+		pass
+
+func test():
+	pass

--- a/modules/gdscript/tests/scripts/analyzer/errors/function_param_type_invalid_contravariance_2.out
+++ b/modules/gdscript/tests/scripts/analyzer/errors/function_param_type_invalid_contravariance_2.out
@@ -1,0 +1,2 @@
+GDTEST_ANALYZER_ERROR
+The function signature doesn't match the parent. Parent signature is "f(Variant) -> Variant".

--- a/modules/gdscript/tests/scripts/analyzer/errors/function_param_type_invalid_contravariance_3.gd
+++ b/modules/gdscript/tests/scripts/analyzer/errors/function_param_type_invalid_contravariance_3.gd
@@ -1,0 +1,10 @@
+class A:
+	func f(_p: int):
+		pass
+
+class B extends A:
+	func f(_p: float): # No implicit conversion.
+		pass
+
+func test():
+	pass

--- a/modules/gdscript/tests/scripts/analyzer/errors/function_param_type_invalid_contravariance_3.out
+++ b/modules/gdscript/tests/scripts/analyzer/errors/function_param_type_invalid_contravariance_3.out
@@ -1,0 +1,2 @@
+GDTEST_ANALYZER_ERROR
+The function signature doesn't match the parent. Parent signature is "f(int) -> Variant".

--- a/modules/gdscript/tests/scripts/analyzer/errors/function_return_type_invalid_covariance_1.gd
+++ b/modules/gdscript/tests/scripts/analyzer/errors/function_return_type_invalid_covariance_1.gd
@@ -1,0 +1,10 @@
+class A:
+	func f() -> Node:
+		return null
+
+class B extends A:
+	func f() -> Object:
+		return null
+
+func test():
+	pass

--- a/modules/gdscript/tests/scripts/analyzer/errors/function_return_type_invalid_covariance_1.out
+++ b/modules/gdscript/tests/scripts/analyzer/errors/function_return_type_invalid_covariance_1.out
@@ -1,0 +1,2 @@
+GDTEST_ANALYZER_ERROR
+The function signature doesn't match the parent. Parent signature is "f() -> Node".

--- a/modules/gdscript/tests/scripts/analyzer/errors/function_return_type_invalid_covariance_2.gd
+++ b/modules/gdscript/tests/scripts/analyzer/errors/function_return_type_invalid_covariance_2.gd
@@ -1,0 +1,10 @@
+class A:
+	func f() -> Node:
+		return null
+
+class B extends A:
+	func f() -> Variant: # No `is_type_compatible()` misuse.
+		return null
+
+func test():
+	pass

--- a/modules/gdscript/tests/scripts/analyzer/errors/function_return_type_invalid_covariance_2.out
+++ b/modules/gdscript/tests/scripts/analyzer/errors/function_return_type_invalid_covariance_2.out
@@ -1,0 +1,2 @@
+GDTEST_ANALYZER_ERROR
+The function signature doesn't match the parent. Parent signature is "f() -> Node".

--- a/modules/gdscript/tests/scripts/analyzer/errors/function_return_type_invalid_covariance_3.gd
+++ b/modules/gdscript/tests/scripts/analyzer/errors/function_return_type_invalid_covariance_3.gd
@@ -1,0 +1,10 @@
+class A:
+	func f() -> Node:
+		return null
+
+class B extends A:
+	func f() -> void: # No `is_type_compatible()` misuse.
+		return
+
+func test():
+	pass

--- a/modules/gdscript/tests/scripts/analyzer/errors/function_return_type_invalid_covariance_3.out
+++ b/modules/gdscript/tests/scripts/analyzer/errors/function_return_type_invalid_covariance_3.out
@@ -1,0 +1,2 @@
+GDTEST_ANALYZER_ERROR
+The function signature doesn't match the parent. Parent signature is "f() -> Node".

--- a/modules/gdscript/tests/scripts/analyzer/errors/function_return_type_invalid_covariance_4.gd
+++ b/modules/gdscript/tests/scripts/analyzer/errors/function_return_type_invalid_covariance_4.gd
@@ -1,0 +1,10 @@
+class A:
+	func f() -> float:
+		return 0.0
+
+class B extends A:
+	func f() -> int: # No implicit conversion.
+		return 0
+
+func test():
+	pass

--- a/modules/gdscript/tests/scripts/analyzer/errors/function_return_type_invalid_covariance_4.out
+++ b/modules/gdscript/tests/scripts/analyzer/errors/function_return_type_invalid_covariance_4.out
@@ -1,0 +1,2 @@
+GDTEST_ANALYZER_ERROR
+The function signature doesn't match the parent. Parent signature is "f() -> float".

--- a/modules/gdscript/tests/scripts/analyzer/features/function_param_type_contravariance.gd
+++ b/modules/gdscript/tests/scripts/analyzer/features/function_param_type_contravariance.gd
@@ -1,0 +1,20 @@
+class A:
+	func int_to_variant(_p: int): pass
+	func node_to_variant(_p: Node): pass
+	func node_2d_to_node(_p: Node2D): pass
+
+	func variant_to_untyped(_p: Variant): pass
+	func int_to_untyped(_p: int): pass
+	func node_to_untyped(_p: Node): pass
+
+class B extends A:
+	func int_to_variant(_p: Variant): pass
+	func node_to_variant(_p: Variant): pass
+	func node_2d_to_node(_p: Node): pass
+
+	func variant_to_untyped(_p): pass
+	func int_to_untyped(_p): pass
+	func node_to_untyped(_p): pass
+
+func test():
+	pass

--- a/modules/gdscript/tests/scripts/analyzer/features/function_param_type_contravariance.out
+++ b/modules/gdscript/tests/scripts/analyzer/features/function_param_type_contravariance.out
@@ -1,0 +1,1 @@
+GDTEST_OK

--- a/modules/gdscript/tests/scripts/analyzer/features/function_return_type_covariance.gd
+++ b/modules/gdscript/tests/scripts/analyzer/features/function_return_type_covariance.gd
@@ -1,0 +1,32 @@
+class A:
+	func variant_to_int() -> Variant: return 0
+	func variant_to_node() -> Variant: return null
+	func node_to_node_2d() -> Node: return null
+
+	func untyped_to_void(): pass
+	func untyped_to_variant(): pass
+	func untyped_to_int(): pass
+	func untyped_to_node(): pass
+
+	func void_to_untyped() -> void: pass
+	func variant_to_untyped() -> Variant: return null
+	func int_to_untyped() -> int: return 0
+	func node_to_untyped() -> Node: return null
+
+class B extends A:
+	func variant_to_int() -> int: return 0
+	func variant_to_node() -> Node: return null
+	func node_to_node_2d() -> Node2D: return null
+
+	func untyped_to_void() -> void: pass
+	func untyped_to_variant() -> Variant: return null
+	func untyped_to_int() -> int: return 0
+	func untyped_to_node() -> Node: return null
+
+	func void_to_untyped(): pass
+	func variant_to_untyped(): pass
+	func int_to_untyped(): pass
+	func node_to_untyped(): pass
+
+func test():
+	pass

--- a/modules/gdscript/tests/scripts/analyzer/features/function_return_type_covariance.out
+++ b/modules/gdscript/tests/scripts/analyzer/features/function_return_type_covariance.out
@@ -1,0 +1,1 @@
+GDTEST_OK

